### PR TITLE
AEAA-363: Vulnerability / Advisory / CVSS data model refactor + public CVSS 2.0/3.1/4.0 implementations

### PR DESCRIPTION
**Pull request by YanWittmann**
This pull request introduces a major refactor of our data models for vulnerabilities and security advisories, along with the public release of our CVSS implementations for versions 2.0, 3.1, and 4.0.

---

## CVSS Standard implementation

Previously, the CVSS Vector management system was closed-source; with this pull request, this will change. This system allows for tracking an arbitrary amount of CVSS vectors on vulnerabilities and security advisories.

This means, the CVSS vectors are not evaluated and merged (potentially conceptually incorrectly) in early stages of the internal vulnerability enrichment process, where not enough information is present to make an informed decision on how to handle such vectors.

The new system evaluates these CVSS vectors only when they are required to be displayed or a score is needed, and evaluates them based on a defined ruleset that can be configured and overwritten by a user. This also allows for displaying the exact origin (source) of the vectors, including links to their hosting entity sites.

In addition to that, the CVSS:4.0 standard has been implemented and is now fully supported in the same ways as the previous versions.

image

image

## Exact tracking of matched vulnerable configurations

The new vulnerability system now fully tracks all possible matching sources that a vulnerability could potentially come from, in addition to the exact vulnerable configuration and artifact that lead to matching the vulnerability.

This makes it possible to display more advanced information in the Vulnerability Assessment Dashboard and the Inventory Report.

image

image

image

image

image

## Separation of Security Advisories from Vulnerabilities

The inventory now no longer saves security advisories in a JSON Array in the Vulnerabilities sheet, but rather in a completely separate sheet Security Advisories. This has multiple advantages:

- Better separation of concerns
- Reducing redundancy of storing the same security advisory potentially many times
- Security advisories are now human-readable

image

## Central Security Policy

Another big change is the introduction of the `CentralSecurityPolicyConfiguration` class as used in the `securityPolicy` parameters. It groups several parameters into a single class. Any properties that are not specified in the XML/JSON configurations are set to their default values.

Please note that any filtering properties from the security configuration only apply at the last steps of the pipelines: the report and the dashboard; but in those, they are applied in the exact same ways.

You can either explicitly set the `securityPolicy` parameters in the report generation and inventory enrichment mojos, or use the `securityPolicyFile` parameter, which will overwrite the `securityPolicy` with the contents of the referenced JSON file.

An example:

```json
{
  "insignificantThreshold": 7.0,
  "includeScoreThreshold": -1.0,
  "includeVulnerabilitiesWithAdvisoryProviders": ["all"],
  "includeAdvisoryTypes": ["all"],
  "vulnerabilityStatusDisplayMapperName": "abstracted",
  "cvssSeverityRanges": "None:pastel-gray:0.0:0.0,Low:strong-yellow:0.1:3.9,Medium:strong-light-orange:4.0:6.9,High:strong-dark-orange:7.0:8.9,Critical:strong-red:9.0:10.0",
  "cvssVersionSelectionPolicy": ["LATEST"],
  "baseCvssSelector": {"stats":[],"rules":[{"method":"ALL","stats":[],"selector":[{"host":["NVD"],"issuerRole":["CNA"],"issuer":["NVD"]},{"host":["Microsoft Corporation"],"issuerRole":["*"],"issuer":["*"]},{"host":["NVD"],"issuerRole":["CNA"],"issuer":["Microsoft Corporation"]},{"host":["GitHub, Inc."],"issuerRole":["*"],"issuer":["*"]},{"host":["NVD"],"issuerRole":["CNA"],"issuer":["GitHub, Inc."]},{"host":["NVD"],"issuerRole":["*"],"issuer":["*"]},{"host":["CERT-SEI"],"issuerRole":["*"],"issuer":["*"]},{"host":["not:Assessment"],"issuerRole":["*"],"issuer":["*"]}],"vectorEval":[]}],"vectorEval":[]},
  "effectiveCvssSelector": {"stats":[{"comparator":"EQUAL","action":"RETURN_NULL","attribute":"assessment","value":0}],"rules":[{"method":"ALL","stats":[],"selector":[{"host":["NVD"],"issuerRole":["CNA"],"issuer":["NVD"]},{"host":["Microsoft Corporation"],"issuerRole":["*"],"issuer":["*"]},{"host":["NVD"],"issuerRole":["CNA"],"issuer":["Microsoft Corporation"]},{"host":["GitHub, Inc."],"issuerRole":["*"],"issuer":["*"]},{"host":["NVD"],"issuerRole":["CNA"],"issuer":["GitHub, Inc."]},{"host":["NVD"],"issuerRole":["*"],"issuer":["*"]},{"host":["CERT-SEI"],"issuerRole":["*"],"issuer":["*"]},{"host":["not:Assessment"],"issuerRole":["*"],"issuer":["*"]}],"vectorEval":[]},{"method":"ALL","stats":[{"provider":"PRESENCE","attribute":"assessment","setType":"ADD"}],"selector":[{"host":["Assessment"],"issuerRole":["*"],"issuer":["all"]}],"vectorEval":[]},{"method":"LOWER","stats":[{"provider":"PRESENCE","attribute":"assessment","setType":"ADD"}],"selector":[{"host":["Assessment"],"issuerRole":["*"],"issuer":["lower"]}],"vectorEval":[]},{"method":"HIGHER","stats":[{"provider":"PRESENCE","attribute":"assessment","setType":"ADD"}],"selector":[{"host":["Assessment"],"issuerRole":["*"],"issuer":["higher"]}],"vectorEval":[]}],"vectorEval":[{"and":["not:IS_BASE_FULLY_DEFINED"],"action":"RETURN_NULL"}]}
}
```

---

These, and many other changes, make handling vulnerabilities, security advisories and CVSS vectors/scores more consistent and more easily (or at all) verifiable. I'm exited to see these changes in action.

### CODE REVIEW

**Comment by YanWittmann** [comment]
> GitHub action build fix attempt

Amazing, it worked

File: libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/reader/AbstractInventoryReader.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import org.apache.poi.ss.usermodel.DataFormatter;
 import org.apache.poi.ss.usermodel.Row;
 import org.metaeffekt.core.inventory.processor.model.*;
+import org.metaeffekt.core.inventory.processor.report.model.aeaa.AeaaInventoryAttribute;
```
Comment (karsten-klein) [inline libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/reader/AbstractInventoryReader.java:24]: Is this import needed. Inventory should not require report.

File: libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/report/model/aeaa/AeaaAmbDataClass.java
Comment (karsten-klein) [inline libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/report/model/aeaa/AeaaAmbDataClass.java:1]: We should consider a better package name and class prefix

File: libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-report-vulnerability/tpc_inventory-cert.dita.vt
Comment (karsten-klein) [inline libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-report-vulnerability/tpc_inventory-cert.dita.vt:1]: security advisors inside text not generally upper case; ok in titles

File: libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-statistics-vulnerability/tpc_inventory-vulnerability-statistics.dita.vt [hunk 2]
```diff
@@ -32,19 +37,22 @@
...
+        <p id="statistics-preface-modified-$advisoryIdentifier">
             The following table shows statistics for the identified vulnerabilities with advisory information from
-            $advisory. The vulnerabilities are included in the statistics with their modified severity if available or
+            $advisoryName. The vulnerabilities are included in the statistics with their modified severity if available or
             their unmodified severity otherwise.
         </p>
-        <p id="statistics-table-modified-$advisory">
-#statisticsOverviewTableModified("vulnerabilities_statistics_table_modified_$advisory.toLowerCase()", "Effective Vulnerability Statistics with $advisory Advisories$reportContext.inContextOf()", "$advisory", $report.getOverviewTablesVulnerabilityStatusMappingFunction())
+        <p id="statistics-table-modified-$advisoryIdentifier">
+#statisticsOverviewTable("vulnerabilities_statistics_table_modified_$advisoryIdentifier.toLowerCase()", "Context Vulnerability Statistics with $advisoryName Advisories$reportContext.inContextOf()", $vulnerabilities, $advisory, true)
```
Comment (karsten-klein) [inline libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-statistics-vulnerability/tpc_inventory-vulnerability-statistics.dita.vt]: Are we sure that we do not need to escape attributes directly embedded in the text.

File: libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/reader/AbstractInventoryReader.java (hunk 1 above)
Comment (YanWittmann) [reply libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/reader/AbstractInventoryReader.java:24]: This import is used for the `AeaaInventoryAttribute.VULNERABILITY_REFERENCED_CONTENT_IDS` enum value for mapping the old way of writing `Referenced Content IDs` to the new one, which is `Referenced Ids`.

The issue with this is that whenever we add an attribute to the corresponding ´AbstractModelBase.Attribute´ enums, they are automatically added to the inventory when writing them out. If you check ´AeaaInventoryAttribute´ or it's Artifact-Analysis counterpart `InventoryAttribute` out, you will see a very large amount of attributes that are used in certain cases, but not always, so adding them to every inventory that is written would lead to much dead space in there.

I would also prefer moving them into the proper abstract model base implementations.

File: libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/report/model/aeaa/AeaaAmbDataClass.java
Comment (YanWittmann) [reply libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/report/model/aeaa/AeaaAmbDataClass.java:1]: I agree. All the data classes copied over share these properties:

- reside in `org.metaeffekt.core.inventory.processor.report.model.aeaa`
- have the exact name as their Artifact Analysis counterpart with the prefix `Aeaa`
- have a doc comment header addition of `Mirrors structure of <code>com.metaeffekt.artifact.analysis.vulnerability.enrichment.AeaaInventoryAttribute</code> until separation of inventory report generation from ae core inventory processor.`

We could / should perform one of these changes:

- change this prefix to something more meaningful
- remove the prefix and replace the indication of a mirrored class by something else

File: libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-statistics-vulnerability/tpc_inventory-vulnerability-statistics.dita.vt (hunk 2 above)
Comment (YanWittmann) [reply libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-statistics-vulnerability/tpc_inventory-vulnerability-statistics.dita.vt]: Do you mean the `advisoryIdentifier` / `advisoryName` in the id and title?

File: libraries/ae-security/src/test/resources/META-INF/LICENSE
Comment (karsten-klein) [inline libraries/ae-security/src/test/resources/META-INF/LICENSE:1]: Why is this required here; licenses are managen on core level automatically. Need to understand.

File: libraries/ae-security/src/main/resources/log4j2.xml
Comment (karsten-klein) [inline libraries/ae-security/src/main/resources/log4j2.xml:1]: log4j is not our default logging implementation. The configuration must not be provided in src/main/resources. Maximum in src/test/resources.

File: libraries/ae-security/src/main/resources/META-INF/LICENSE
Comment (karsten-klein) [inline libraries/ae-security/src/main/resources/META-INF/LICENSE:1]: License should be managed automatically. Need to understand.

File: libraries/ae-security/.gitignore
Comment (karsten-klein) [inline libraries/ae-security/.gitignore:1]: A local .gitignore should not be required.

File: libraries/ae-security/pom.xml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+            <optional>true</optional>
```
Comment (karsten-klein) [inline libraries/ae-security/pom.xml:62]: Change scope to test for log4j dependencies

**Review by karsten-klein — COMMENTED** [review]
Hi Yan,

we need to at least add javadoc on class level. Also the setup regarding the license header plugin, license handling and logging configuration must be revised.

File: libraries/ae-security/src/main/resources/META-INF/LICENSE
Comment (YanWittmann) [reply libraries/ae-security/src/main/resources/META-INF/LICENSE:1]: The LICENSE file is automatically created when I build the module in those two directories. I did not add them in the main/resources/META-INF/LICENSE and test/resources/META-INF/LICENSE. It seems as though it is added by the `<artifactId>license-maven-plugin</artifactId>` plugin.

File: libraries/ae-security/src/main/resources/log4j2.xml
Comment (YanWittmann) [reply libraries/ae-security/src/main/resources/log4j2.xml:1]: Moved to src/test/resources.

File: libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-statistics-vulnerability/tpc_inventory-vulnerability-statistics.dita.vt (hunk 2 above)
Comment (YanWittmann) [reply libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-statistics-vulnerability/tpc_inventory-vulnerability-statistics.dita.vt]: All vulnerability and advisory names are now escaped when used in the id attribute or in title tags.

File: libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-report-vulnerability/tpc_inventory-cert.dita.vt
Comment (YanWittmann) [reply libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-report-vulnerability/tpc_inventory-cert.dita.vt:1]: Changed to lowercase in all text except titles.

**Comment by YanWittmann** [comment]
... it's been a long time coming!
